File: streetlab-backend/schema.py
"""StreetLab wire schema — the Python mirror of ``streetlab/src/schema.ts``.

The TypeScript file is the single source of truth. This module is a hand-written
pydantic v2 transcription of it, laid out in the same order and with the same
section headings so the two can be read side by side. Field names are verbatim,
including ``type`` and ``cls``, which shadow Python conventions but are what the
wire carries.

Conventions (unchanged from the TypeScript)
  - Distances in metres, speeds in m/s, accelerations in m/s^2.
  - Angles in radians. Headings are 0 at +x (east) and increase CCW.
  - World is a right-handed 2D plane: +x east, +y north.
  - Time ``t`` is simulator seconds since scenario start.

Three transcription hazards are handled deliberately; see the tests that pin them:

  - ``z.number()`` rejects NaN and Infinity, but pydantic allows both by default.
    Every numeric field therefore goes through the ``Num`` family of aliases,
    which set ``allow_inf_nan=False``. One non-finite float would otherwise fail
    zod validation on arrival and silently freeze the frontend.
  - ``.nullable()`` in zod means "present, possibly null" — not "may be absent".
    Never serialise with ``exclude_none=True``.
  - zod strips unknown keys rather than rejecting them, so a stray Python-only
    field is invisible to the frontend. The fixture round-trip tests compare full
    dictionaries, which catches additions and omissions alike.
"""

from typing import Annotated, Any, Generic, Literal, TypeVar, Union

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter, ValidationError

# The wire protocol version, mirroring PROTOCOL_VERSION in schema.ts. Every
# message carries it in a field named `protocol`.
PROTOCOL_VERSION = 2

# This Python package's own version. Deliberately distinct from the wire
# protocol and never serialised — the two version independently.
SCHEMA_VERSION = "0.1.0"

# --------------------------------------------------------------------------- #
# Primitives                                                                   #
# --------------------------------------------------------------------------- #

# `z.number()` — finite only.
Num = Annotated[float, Field(allow_inf_nan=False)]
NonNeg = Annotated[float, Field(ge=0, allow_inf_nan=False)]
Pos = Annotated[float, Field(gt=0, allow_inf_nan=False)]
# `z.number().min(0).max(1)`
Unit = Annotated[float, Field(ge=0, le=1, allow_inf_nan=False)]

# `[x, y]` in world metres.
Vec2 = tuple[Num, Num]

HexColor = Annotated[str, Field(pattern=r"^#[0-9a-fA-F]{6}$")]


class Wire(BaseModel):
    """Base for every wire model. Unknown keys are ignored, matching zod's strip."""

    model_config = ConfigDict(extra="ignore")


class Pose(Wire):
    x: Num
    y: Num
    # radians, 0 = +x (east), CCW positive
    heading: Num


class Size(Wire):
    length: Pos
    width: Pos
    height: Pos


# --------------------------------------------------------------------------- #
# SceneDescription — static world, sent once per scenario load                  #
# --------------------------------------------------------------------------- #

SignalPhase = Literal["red", "yellow", "green", "flashing_yellow", "off"]
RoadClass = Literal["arterial", "collector", "residential", "service"]
LaneMarking = Literal["none", "dashed_white", "solid_white", "double_yellow"]


class Road(Wire):
    id: str
    name: str
    road_class: RoadClass
    # Ordered polyline down the middle of the carriageway.
    centerline: Annotated[list[Vec2], Field(min_length=2)]
    lanes_forward: Annotated[int, Field(ge=0)]
    lanes_backward: Annotated[int, Field(ge=0)]
    lane_width_m: Pos
    speed_limit_mps: NonNeg
    oneway: bool
    # Marking drawn on the centre divider.
    center_marking: LaneMarking
    has_sidewalk: bool


class Building(Wire):
    id: str
    # CCW footprint ring; not closed (first point is not repeated).
    footprint: Annotated[list[Vec2], Field(min_length=3)]
    height_m: Pos
    color: HexColor
    roof_color: HexColor


class Crosswalk(Wire):
    id: str
    center: Vec2
    # Direction pedestrians walk, radians.
    heading: Num
    # Depth of the striped band, measured perpendicular to `heading`.
    width_m: Pos
    # Extent along `heading` — i.e. the carriageway width being crossed.
    length_m: Pos
    style: Literal["ladder", "continental", "transverse"]


class TrafficLight(Wire):
    id: str
    position: Vec2
    # Direction the lamp faces — i.e. toward the traffic it governs.
    heading: Num
    # Horizontal mast-arm reach; 0 means a pole-mounted head.
    mast_arm_m: NonNeg
    height_m: Pos


class StopSign(Wire):
    id: str
    position: Vec2
    heading: Num


class Tree(Wire):
    id: str
    position: Vec2
    height_m: Pos
    canopy_radius_m: Pos
    trunk_radius_m: Pos
    # 0..1 hue jitter so a forest of clones does not look like one.
    variant: Unit


class StreetSign(Wire):
    id: str
    position: Vec2
    heading: Num
    text: str
    kind: Literal["street_name", "speed_limit", "no_parking"]


class ScenarioSummary(Wire):
    """One entry in the left sidebar's saved-scenario list."""

    id: str
    # 1-based display index, rendered as "01".."05".
    index: Annotated[int, Field(gt=0)]
    name: str
    location: str
    description: str
    duration_s: NonNeg
    bookmarked: bool
    difficulty: Literal["easy", "moderate", "hard"]
    # Road skeleton for the minimap thumbnail, in thumbnail-local units.
    preview_paths: list[list[Vec2]]
    # Ego route for the minimap thumbnail.
    preview_route: list[Vec2]


class Origin(Wire):
    lat: Num
    lon: Num


class Bounds(Wire):
    min_x: Num
    min_y: Num
    max_x: Num
    max_y: Num


class SceneDescription(Wire):
    type: Literal["scene_description"] = "scene_description"
    protocol: int = PROTOCOL_VERSION
    scene_id: str
    scenario_id: str
    name: str
    # Human-readable neighbourhood, e.g. "Nob Hill".
    location: str
    # ODbL requires crediting OpenStreetMap wherever its data is shown.
    attribution: str
    origin: Origin
    bounds: Bounds
    roads: list[Road]
    buildings: list[Building]
    crosswalks: list[Crosswalk]
    traffic_lights: list[TrafficLight]
    stop_signs: list[StopSign]
    trees: list[Tree]
    street_signs: list[StreetSign]
    # Scenarios the server can load; drives the left sidebar.
    catalog: list[ScenarioSummary]


# --------------------------------------------------------------------------- #
# StateUpdate — streamed at sim rate                                           #
# --------------------------------------------------------------------------- #

DetectionClass = Literal[
    "car", "truck", "bus", "motorcycle", "cyclist", "pedestrian", "unknown"
]


class Detection(Wire):
    id: str
    cls: DetectionClass
    pose: Pose
    size: Size
    # World-frame velocity `[vx, vy]` in m/s.
    velocity: Vec2
    speed_mps: Num
    confidence: Unit
    hazard: bool
    # Shown on the 3D billboard when `hazard` is true.
    hazard_label: str | None
    ttc_s: Num | None
    # Lane index relative to ego: -1 right, 0 same, +1 left, null if unknown.
    lane_offset: int | None


class RadarPoint(Wire):
    id: str | None
    # Ego-frame bearing, radians. 0 = dead ahead, + = left.
    azimuth: Num
    range_m: NonNeg
    # Closing rate; negative means approaching.
    range_rate_mps: Num
    rcs_db: Num
    # True for points associated with a tracked detection (vs clutter).
    tracked: bool


class LaneNeighbor(Wire):
    id: str
    cls: DetectionClass
    # -1 right, 0 same lane, +1 left.
    lane_offset: int
    # Signed distance ahead of ego along the lane.
    longitudinal_m: Num
    # Signed lateral offset from ego's lane centre, + = left.
    lateral_m: Num
    speed_mps: Num
    hazard: bool


class LaneState(Wire):
    lane_index: int
    lane_count: Annotated[int, Field(gt=0)]
    lane_width_m: Pos
    # Ego's signed lateral offset from its lane centre, + = left.
    offset_m: Num
    # Ego heading minus lane heading, radians.
    heading_error: Num
    left_marking: LaneMarking
    right_marking: LaneMarking
    neighbors: list[LaneNeighbor]


class Subsystem(Wire):
    key: str
    label: str
    status: Literal["ok", "warn", "fault"]
    detail: str | None


class VehicleStatus(Wire):
    battery_pct: Annotated[float, Field(ge=0, le=100, allow_inf_nan=False)]
    range_km: NonNeg
    motor_temp_c: Num
    # [FL, FR, RL, RR] in kPa.
    tire_pressure_kpa: tuple[Num, Num, Num, Num]
    subsystems: list[Subsystem]
    overall: Literal["ok", "warn", "fault"]


class TrajectorySample(Wire):
    # Seconds relative to this frame; negative is observed history.
    t: Num
    # Lateral offset from the current lane centre, + = left.
    lateral_m: Num


class TrajectoryPrediction(Wire):
    horizon_s: Pos
    planned: list[TrajectorySample]
    # Predicted path of the cutting-in agent, or null when nobody is cutting in.
    cutin: list[TrajectorySample] | None
    cutin_label: str | None


class Telemetry(Wire):
    radar: list[RadarPoint]
    lane: LaneState
    # Time-to-collision with the most critical object, seconds.
    ttc_s: Num | None
    vehicle: VehicleStatus
    trajectory: TrajectoryPrediction


Maneuver = Literal[
    "keep_lane",
    "turn_left",
    "turn_right",
    "lane_change_left",
    "lane_change_right",
    "stop",
    "yield",
]


class Plan(Wire):
    # Forward plan in world coordinates, ego-first.
    polyline: list[Vec2]
    target_speed_mps: NonNeg
    maneuver: Maneuver
    confidence: Unit


CruiseMode = Literal["off", "cruise", "autosteer", "fsd"]


class Cruise(Wire):
    mode: CruiseMode
    set_speed_mps: NonNeg


class Ego(Wire):
    pose: Pose
    speed_mps: Num
    accel_mps2: Num
    # Road-wheel angle in radians, + = left.
    steering_angle: Num
    yaw_rate: Num
    throttle: Unit
    brake: Unit
    gear: Literal["P", "R", "N", "D"]
    speed_limit_mps: NonNeg
    cruise: Cruise
    size: Size


class SignalState(Wire):
    id: str
    phase: SignalPhase
    time_to_change_s: Num | None


class SimEvent(Wire):
    t: Num
    level: Literal["info", "warn", "critical"]
    code: str
    message: str


class StateUpdate(Wire):
    type: Literal["state_update"] = "state_update"
    protocol: int = PROTOCOL_VERSION
    seq: Annotated[int, Field(ge=0)]
    t: NonNeg
    sim_rate_hz: Pos
    paused: bool
    assist_active: bool
    scenario_id: str
    ego: Ego
    detections: list[Detection]
    plan: Plan
    telemetry: Telemetry
    signals: list[SignalState]
    events: list[SimEvent]


# --------------------------------------------------------------------------- #
# Command — client -> server                                                   #
# --------------------------------------------------------------------------- #

LayerKey = Literal[
    "detections",
    "plan_path",
    "lane_markings",
    "crosswalks",
    "buildings",
    "trees",
    "traffic_lights",
    "radar_cone",
    "labels",
]

CameraView = Literal["chase", "overhead", "cockpit", "free"]

# `z.union([z.number(), z.string(), z.boolean()])`. `bool` precedes the numeric
# types because Python's bool is a subclass of int; pydantic's smart union
# matches exact types first, and `int` precedes `float` so an integral value
# survives the round trip as an integer rather than becoming 35.0.
ParamValue = Union[bool, int, float, str]


class _Cmd(Wire):
    """Every command carries a client-generated id so an Ack can be correlated."""

    id: str


class SetPaused(_Cmd):
    cmd: Literal["set_paused"] = "set_paused"
    paused: bool


class Step(_Cmd):
    cmd: Literal["step"] = "step"
    frames: Annotated[int, Field(gt=0)]


class Reset(_Cmd):
    cmd: Literal["reset"] = "reset"


class LoadScenario(_Cmd):
    cmd: Literal["load_scenario"] = "load_scenario"
    scenario_id: str


class LoadLocation(_Cmd):
    cmd: Literal["load_location"] = "load_location"
    query: Annotated[str, Field(min_length=1)]
    # Absent means "use the location's default". zod `.optional()` allows the
    # key to be missing, unlike `.nullable()` which would require it present.
    radius_m: Pos | None = None


class SetParam(_Cmd):
    cmd: Literal["set_param"] = "set_param"
    key: str
    value: ParamValue


class ToggleLayer(_Cmd):
    cmd: Literal["toggle_layer"] = "toggle_layer"
    layer: LayerKey
    visible: bool


class SetCamera(_Cmd):
    cmd: Literal["set_camera"] = "set_camera"
    view: CameraView


class InjectHazard(_Cmd):
    cmd: Literal["inject_hazard"] = "inject_hazard"
    kind: str


Command = Annotated[
    Union[
        SetPaused,
        Step,
        Reset,
        LoadScenario,
        LoadLocation,
        SetParam,
        ToggleLayer,
        SetCamera,
        InjectHazard,
    ],
    Field(discriminator="cmd"),
]


# --------------------------------------------------------------------------- #
# Ack — server -> client                                                       #
# --------------------------------------------------------------------------- #


class Ack(Wire):
    type: Literal["ack"] = "ack"
    protocol: int = PROTOCOL_VERSION
    # Echoes `Command.id`.
    id: str
    cmd: str
    ok: bool
    message: str | None
    t: Num


# --------------------------------------------------------------------------- #
# Envelope + helpers                                                           #
# --------------------------------------------------------------------------- #

ServerMessage = Annotated[
    Union[SceneDescription, StateUpdate, Ack], Field(discriminator="type")
]

_COMMAND_ADAPTER: TypeAdapter[Any] = TypeAdapter(Command)
_SERVER_MESSAGE_ADAPTER: TypeAdapter[Any] = TypeAdapter(ServerMessage)

T = TypeVar("T")


class ParseResult(BaseModel, Generic[T]):
    """Mirror of the TypeScript `ParseResult<T>` discriminated result."""

    model_config = ConfigDict(arbitrary_types_allowed=True)

    ok: bool
    value: T | None = None
    error: str | None = None


def format_issues(err: ValidationError) -> str:
    """Format a validation failure into a single compact line for logs."""
    return "; ".join(
        f"{'.'.join(str(p) for p in issue['loc']) or '<root>'}: {issue['msg']}"
        for issue in err.errors()[:4]
    )


def parse_server_message(raw: Any) -> ParseResult:
    """Validate an outbound server message. Never raises."""
    try:
        return ParseResult(ok=True, value=_SERVER_MESSAGE_ADAPTER.validate_python(raw))
    except ValidationError as err:
        return ParseResult(ok=False, error=format_issues(err))


def parse_command(raw: Any) -> ParseResult:
    """Validate an inbound command. Never raises — a malformed command from the
    wire must degrade to a logged warning and a failed ack, never a crash."""
    try:
        return ParseResult(ok=True, value=_COMMAND_ADAPTER.validate_python(raw))
    except ValidationError as err:
        return ParseResult(ok=False, error=format_issues(err))


# All layer keys, in the order the Layers tab should present them.
LAYER_KEYS: tuple[str, ...] = (
    "detections",
    "plan_path",
    "lane_markings",
    "crosswalks",
    "buildings",
    "trees",
    "traffic_lights",
    "radar_cone",
    "labels",
)

File: streetlab/src/schema.ts
/**
 * StreetLab wire schema.
 *
 * This file is the single source of truth for the frontend<->simulator
 * protocol. The backend (a separate Python process) will emit exactly these
 * messages over a WebSocket; the in-process mock emits the same shapes so the
 * UI never learns whether it is talking to a real simulator or not.
 *
 * Conventions
 *   - Distances in metres, speeds in m/s, accelerations in m/s^2.
 *   - Angles in radians. Headings are 0 at +x (east) and increase CCW.
 *   - World is a right-handed 2D plane: +x east, +y north. The renderer maps
 *     this to three.js as (x, 0, -y) so that +y (north) points at -z.
 *   - Time `t` is simulator seconds since scenario start.
 */
import { z } from 'zod';

export const PROTOCOL_VERSION = 3;

/* ------------------------------------------------------------------ */
/* Primitives                                                          */
/* ------------------------------------------------------------------ */

/** `[x, y]` in world metres. Tuples keep 60 Hz polyline payloads compact. */
export const Vec2Schema = z.tuple([z.number(), z.number()]);

export const PoseSchema = z.object({
  x: z.number(),
  y: z.number(),
  /** radians, 0 = +x (east), CCW positive */
  heading: z.number(),
});

export const SizeSchema = z.object({
  length: z.number().positive(),
  width: z.number().positive(),
  height: z.number().positive(),
});

const HexColorSchema = z.string().regex(/^#[0-9a-fA-F]{6}$/, 'expected #rrggbb');

/* ------------------------------------------------------------------ */
/* SceneDescription — static world, sent once per scenario load        */
/* ------------------------------------------------------------------ */

export const SignalPhaseSchema = z.enum([
  'red',
  'yellow',
  'green',
  'flashing_yellow',
  'off',
]);

export const RoadClassSchema = z.enum([
  'arterial',
  'collector',
  'residential',
  'service',
]);

export const LaneMarkingSchema = z.enum([
  'none',
  'dashed_white',
  'solid_white',
  'double_yellow',
]);

export const RoadSchema = z.object({
  id: z.string(),
  name: z.string(),
  road_class: RoadClassSchema,
  /** Ordered polyline down the middle of the carriageway. */
  centerline: z.array(Vec2Schema).min(2),
  lanes_forward: z.number().int().min(0),
  lanes_backward: z.number().int().min(0),
  lane_width_m: z.number().positive(),
  speed_limit_mps: z.number().nonnegative(),
  oneway: z.boolean(),
  /** Marking drawn on the centre divider. */
  center_marking: LaneMarkingSchema,
  has_sidewalk: z.boolean(),
});

export const BuildingSchema = z.object({
  id: z.string(),
  /** CCW footprint ring; not closed (first point is not repeated). */
  footprint: z.array(Vec2Schema).min(3),
  height_m: z.number().positive(),
  color: HexColorSchema,
  roof_color: HexColorSchema,
});

export const CrosswalkSchema = z.object({
  id: z.string(),
  center: Vec2Schema,
  /** Direction pedestrians walk, radians. */
  heading: z.number(),
  /** Depth of the striped band, measured perpendicular to `heading`. */
  width_m: z.number().positive(),
  /** Extent along `heading` — i.e. the carriageway width being crossed. */
  length_m: z.number().positive(),
  style: z.enum(['ladder', 'continental', 'transverse']),
});

export const TrafficLightSchema = z.object({
  id: z.string(),
  position: Vec2Schema,
  /** Direction the lamp faces — i.e. toward the traffic it governs. */
  heading: z.number(),
  /** Horizontal mast-arm reach; 0 means a pole-mounted head. */
  mast_arm_m: z.number().nonnegative(),
  height_m: z.number().positive(),
});

export const StopSignSchema = z.object({
  id: z.string(),
  position: Vec2Schema,
  heading: z.number(),
});

export const TreeSchema = z.object({
  id: z.string(),
  position: Vec2Schema,
  height_m: z.number().positive(),
  canopy_radius_m: z.number().positive(),
  trunk_radius_m: z.number().positive(),
  /** 0..1 hue jitter so a forest of clones does not look like one. */
  variant: z.number().min(0).max(1),
});

export const StreetSignSchema = z.object({
  id: z.string(),
  position: Vec2Schema,
  heading: z.number(),
  text: z.string(),
  kind: z.enum(['street_name', 'speed_limit', 'no_parking']),
});

/** One entry in the left sidebar's saved-scenario list. */
export const ScenarioSummarySchema = z.object({
  id: z.string(),
  /** 1-based display index, rendered as "01".."05". */
  index: z.number().int().positive(),
  name: z.string(),
  location: z.string(),
  description: z.string(),
  duration_s: z.number().nonnegative(),
  bookmarked: z.boolean(),
  difficulty: z.enum(['easy', 'moderate', 'hard']),
  /** Road skeleton for the minimap thumbnail, in thumbnail-local units. */
  preview_paths: z.array(z.array(Vec2Schema)),
  /** Ego route for the minimap thumbnail. */
  preview_route: z.array(Vec2Schema),
});

export const SceneDescriptionSchema = z.object({
  type: z.literal('scene_description'),
  protocol: z.number().int(),
  scene_id: z.string(),
  scenario_id: z.string(),
  name: z.string(),
  /** Human-readable neighbourhood, e.g. "Nob Hill". */
  location: z.string(),
  // ODbL requires crediting OpenStreetMap wherever its data is shown.
  attribution: z.string(),
  origin: z.object({ lat: z.number(), lon: z.number() }),
  bounds: z.object({
    min_x: z.number(),
    min_y: z.number(),
    max_x: z.number(),
    max_y: z.number(),
  }),
  roads: z.array(RoadSchema),
  buildings: z.array(BuildingSchema),
  crosswalks: z.array(CrosswalkSchema),
  traffic_lights: z.array(TrafficLightSchema),
  stop_signs: z.array(StopSignSchema),
  trees: z.array(TreeSchema),
  street_signs: z.array(StreetSignSchema),
  /** Scenarios the server can load; drives the left sidebar. */
  catalog: z.array(ScenarioSummarySchema),
});

/* ------------------------------------------------------------------ */
/* StateUpdate — streamed at sim rate                                  */
/* ------------------------------------------------------------------ */

export const DetectionClassSchema = z.enum([
  'car',
  'truck',
  'bus',
  'motorcycle',
  'cyclist',
  'pedestrian',
  'unknown',
]);

export const DetectionSchema = z.object({
  id: z.string(),
  cls: DetectionClassSchema,
  pose: PoseSchema,
  size: SizeSchema,
  /** World-frame velocity `[vx, vy]` in m/s. */
  velocity: Vec2Schema,
  speed_mps: z.number(),
  confidence: z.number().min(0).max(1),
  hazard: z.boolean(),
  /** Shown on the 3D billboard when `hazard` is true. */
  hazard_label: z.string().nullable(),
  ttc_s: z.number().nullable(),
  /** Lane index relative to ego: -1 right, 0 same, +1 left, null if unknown. */
  lane_offset: z.number().int().nullable(),
});

export const PerceptionModeSchema = z.enum(['ground-truth', 'ml']);

/**
 * The camera that produced one frame, in WIRE world coordinates:
 * `+x` east, `+y` north, `+z` up, ground plane at `z = 0`.
 * The frontend converts out of Three.js's Y-up frame before sending, so the
 * backend never learns that a renderer convention exists.
 */
export const CameraParamsSchema = z.object({
  x: z.number(),
  y: z.number(),
  z: z.number(),
  /** radians, 0 = +x (east), CCW positive — same convention as Pose.heading */
  yaw: z.number(),
  pitch: z.number(),
  roll: z.number(),
  fov_y_deg: z.number().positive(),
  aspect: z.number().positive(),
});

/** Transport and quality numbers for the ML perception path. */
export const PerceptionStatsSchema = z.object({
  mode: PerceptionModeSchema,
  /** Model inference time. Null until Phase 2 lands a model. */
  detector_ms: z.number().nonnegative().nullable(),
  /** Frame render -> detections available. */
  e2e_ms: z.number().nonnegative().nullable(),
  frames_received: z.number().int().nonnegative(),
  frames_dropped: z.number().int().nonnegative(),
  /** Quality fields stay null until scoring lands in Phase 3. */
  precision: z.number().min(0).max(1).nullable(),
  recall: z.number().min(0).max(1).nullable(),
  mean_pos_err_m: z.number().nonnegative().nullable(),
});

export const RadarPointSchema = z.object({
  id: z.string().nullable(),
  /** Ego-frame bearing, radians. 0 = dead ahead, + = left. */
  azimuth: z.number(),
  range_m: z.number().nonnegative(),
  /** Closing rate; negative means approaching. */
  range_rate_mps: z.number(),
  rcs_db: z.number(),
  /** True for points associated with a tracked detection (vs clutter). */
  tracked: z.boolean(),
});

export const LaneNeighborSchema = z.object({
  id: z.string(),
  cls: DetectionClassSchema,
  /** -1 right, 0 same lane, +1 left. */
  lane_offset: z.number().int(),
  /** Signed distance ahead of ego along the lane. */
  longitudinal_m: z.number(),
  /** Signed lateral offset from ego's lane centre, + = left. */
  lateral_m: z.number(),
  speed_mps: z.number(),
  hazard: z.boolean(),
});

export const LaneStateSchema = z.object({
  lane_index: z.number().int(),
  lane_count: z.number().int().positive(),
  lane_width_m: z.number().positive(),
  /** Ego's signed lateral offset from its lane centre, + = left. */
  offset_m: z.number(),
  /** Ego heading minus lane heading, radians. */
  heading_error: z.number(),
  left_marking: LaneMarkingSchema,
  right_marking: LaneMarkingSchema,
  neighbors: z.array(LaneNeighborSchema),
});

export const SubsystemSchema = z.object({
  key: z.string(),
  label: z.string(),
  status: z.enum(['ok', 'warn', 'fault']),
  detail: z.string().nullable(),
});

export const VehicleStatusSchema = z.object({
  battery_pct: z.number().min(0).max(100),
  range_km: z.number().nonnegative(),
  motor_temp_c: z.number(),
  /** [FL, FR, RL, RR] in kPa. */
  tire_pressure_kpa: z.tuple([
    z.number(),
    z.number(),
    z.number(),
    z.number(),
  ]),
  subsystems: z.array(SubsystemSchema),
  overall: z.enum(['ok', 'warn', 'fault']),
});

export const TrajectorySampleSchema = z.object({
  /** Seconds relative to this frame; negative is observed history. */
  t: z.number(),
  /** Lateral offset from the current lane centre, + = left. */
  lateral_m: z.number(),
});

export const TrajectoryPredictionSchema = z.object({
  horizon_s: z.number().positive(),
  planned: z.array(TrajectorySampleSchema),
  /** Predicted path of the cutting-in agent, or null when nobody is cutting in. */
  cutin: z.array(TrajectorySampleSchema).nullable(),
  cutin_label: z.string().nullable(),
});

export const TelemetrySchema = z.object({
  radar: z.array(RadarPointSchema),
  lane: LaneStateSchema,
  /** Time-to-collision with the most critical object, seconds. */
  ttc_s: z.number().nullable(),
  vehicle: VehicleStatusSchema,
  trajectory: TrajectoryPredictionSchema,
});

export const ManeuverSchema = z.enum([
  'keep_lane',
  'turn_left',
  'turn_right',
  'lane_change_left',
  'lane_change_right',
  'stop',
  'yield',
]);

export const PlanSchema = z.object({
  /** Forward plan in world coordinates, ego-first. */
  polyline: z.array(Vec2Schema),
  target_speed_mps: z.number().nonnegative(),
  maneuver: ManeuverSchema,
  confidence: z.number().min(0).max(1),
});

export const CruiseModeSchema = z.enum(['off', 'cruise', 'autosteer', 'fsd']);

export const EgoSchema = z.object({
  pose: PoseSchema,
  speed_mps: z.number(),
  accel_mps2: z.number(),
  /** Road-wheel angle in radians, + = left. */
  steering_angle: z.number(),
  yaw_rate: z.number(),
  throttle: z.number().min(0).max(1),
  brake: z.number().min(0).max(1),
  gear: z.enum(['P', 'R', 'N', 'D']),
  speed_limit_mps: z.number().nonnegative(),
  cruise: z.object({
    mode: CruiseModeSchema,
    set_speed_mps: z.number().nonnegative(),
  }),
  size: SizeSchema,
});

export const SignalStateSchema = z.object({
  id: z.string(),
  phase: SignalPhaseSchema,
  time_to_change_s: z.number().nullable(),
});

export const SimEventSchema = z.object({
  t: z.number(),
  level: z.enum(['info', 'warn', 'critical']),
  code: z.string(),
  message: z.string(),
});

export const StateUpdateSchema = z.object({
  type: z.literal('state_update'),
  protocol: z.number().int(),
  seq: z.number().int().nonnegative(),
  t: z.number().nonnegative(),
  sim_rate_hz: z.number().positive(),
  paused: z.boolean(),
  assist_active: z.boolean(),
  scenario_id: z.string(),
  ego: EgoSchema,
  detections: z.array(DetectionSchema),
  plan: PlanSchema,
  telemetry: TelemetrySchema,
  signals: z.array(SignalStateSchema),
  events: z.array(SimEventSchema),
  /** Null when no ML perception is running — distinct from "measured, and zero". */
  perception: PerceptionStatsSchema.nullable(),
});

/* ------------------------------------------------------------------ */
/* Command — client -> server                                          */
/* ------------------------------------------------------------------ */

export const LayerKeySchema = z.enum([
  'detections',
  'plan_path',
  'lane_markings',
  'crosswalks',
  'buildings',
  'trees',
  'traffic_lights',
  'radar_cone',
  'labels',
]);

export const CameraViewSchema = z.enum(['chase', 'overhead', 'cockpit', 'free']);

export const ParamValueSchema = z.union([
  z.number(),
  z.string(),
  z.boolean(),
]);

/** Every command carries a client-generated id so an Ack can be correlated. */
const cmd = <S extends z.ZodRawShape>(shape: S) =>
  z.object({ id: z.string(), ...shape });

export const CommandSchema = z.discriminatedUnion('cmd', [
  cmd({ cmd: z.literal('set_paused'), paused: z.boolean() }),
  cmd({ cmd: z.literal('step'), frames: z.number().int().positive() }),
  cmd({ cmd: z.literal('reset') }),
  cmd({ cmd: z.literal('load_scenario'), scenario_id: z.string() }),
  cmd({
    cmd: z.literal('load_location'),
    query: z.string().min(1),
    radius_m: z.number().positive().optional(),
  }),
  cmd({
    cmd: z.literal('set_param'),
    key: z.string(),
    value: ParamValueSchema,
  }),
  cmd({
    cmd: z.literal('toggle_layer'),
    layer: LayerKeySchema,
    visible: z.boolean(),
  }),
  cmd({ cmd: z.literal('set_camera'), view: CameraViewSchema }),
  cmd({ cmd: z.literal('inject_hazard'), kind: z.string() }),
  cmd({ cmd: z.literal('set_perception'), mode: PerceptionModeSchema }),
  cmd({
    cmd: z.literal('camera_frame'),
    /** Monotonic per connection; the backend drops anything out of order. */
    seq: z.number().int().nonnegative(),
    /** Sim seconds the frame depicts. */
    t: z.number(),
    width: z.number().int().positive(),
    height: z.number().int().positive(),
    format: z.literal('jpeg'),
    /** base64. Capped: an uncapped field here is an OOM waiting for a bad client. */
    data: z.string().max(524288),
    camera: CameraParamsSchema,
  }),
]);

/* ------------------------------------------------------------------ */
/* Ack — server -> client                                              */
/* ------------------------------------------------------------------ */

export const AckSchema = z.object({
  type: z.literal('ack'),
  protocol: z.number().int(),
  /** Echoes `Command.id`. */
  id: z.string(),
  cmd: z.string(),
  ok: z.boolean(),
  message: z.string().nullable(),
  t: z.number(),
});

/* ------------------------------------------------------------------ */
/* Envelope + helpers                                                  */
/* ------------------------------------------------------------------ */

export const ServerMessageSchema = z.discriminatedUnion('type', [
  SceneDescriptionSchema,
  StateUpdateSchema,
  AckSchema,
]);

/* ---- inferred types ---- */

export type Vec2 = z.infer<typeof Vec2Schema>;
export type Pose = z.infer<typeof PoseSchema>;
export type Size = z.infer<typeof SizeSchema>;
export type SignalPhase = z.infer<typeof SignalPhaseSchema>;
export type RoadClass = z.infer<typeof RoadClassSchema>;
export type LaneMarking = z.infer<typeof LaneMarkingSchema>;
export type Road = z.infer<typeof RoadSchema>;
export type Building = z.infer<typeof BuildingSchema>;
export type Crosswalk = z.infer<typeof CrosswalkSchema>;
export type TrafficLight = z.infer<typeof TrafficLightSchema>;
export type StopSign = z.infer<typeof StopSignSchema>;
export type Tree = z.infer<typeof TreeSchema>;
export type StreetSign = z.infer<typeof StreetSignSchema>;
export type ScenarioSummary = z.infer<typeof ScenarioSummarySchema>;
export type SceneDescription = z.infer<typeof SceneDescriptionSchema>;

export type DetectionClass = z.infer<typeof DetectionClassSchema>;
export type Detection = z.infer<typeof DetectionSchema>;
export type PerceptionMode = z.infer<typeof PerceptionModeSchema>;
export type CameraParams = z.infer<typeof CameraParamsSchema>;
export type PerceptionStats = z.infer<typeof PerceptionStatsSchema>;
export type RadarPoint = z.infer<typeof RadarPointSchema>;
export type LaneNeighbor = z.infer<typeof LaneNeighborSchema>;
export type LaneState = z.infer<typeof LaneStateSchema>;
export type Subsystem = z.infer<typeof SubsystemSchema>;
export type VehicleStatus = z.infer<typeof VehicleStatusSchema>;
export type TrajectorySample = z.infer<typeof TrajectorySampleSchema>;
export type TrajectoryPrediction = z.infer<typeof TrajectoryPredictionSchema>;
export type Telemetry = z.infer<typeof TelemetrySchema>;
export type Maneuver = z.infer<typeof ManeuverSchema>;
export type Plan = z.infer<typeof PlanSchema>;
export type CruiseMode = z.infer<typeof CruiseModeSchema>;
export type Ego = z.infer<typeof EgoSchema>;
export type SignalState = z.infer<typeof SignalStateSchema>;
export type SimEvent = z.infer<typeof SimEventSchema>;
export type StateUpdate = z.infer<typeof StateUpdateSchema>;

export type LayerKey = z.infer<typeof LayerKeySchema>;
export type CameraView = z.infer<typeof CameraViewSchema>;
export type ParamValue = z.infer<typeof ParamValueSchema>;
export type Command = z.infer<typeof CommandSchema>;

/**
 * A command as callers write it: the correlation `id` is filled in by the store,
 * so UI code only supplies the payload. Distributes over the union so each
 * member keeps its own required fields.
 */
export type CommandInput = Command extends infer C
  ? C extends { id: string }
    ? Omit<C, 'id'> & { id?: string }
    : never
  : never;
export type Ack = z.infer<typeof AckSchema>;
export type ServerMessage = z.infer<typeof ServerMessageSchema>;

export type ParseResult<T> =
  | { ok: true; value: T }
  | { ok: false; error: string };

/** Format a zod failure into a single compact line for logs/UI. */
export function formatIssues(err: z.ZodError): string {
  return err.issues
    .slice(0, 4)
    .map((i) => `${i.path.join('.') || '<root>'}: ${i.message}`)
    .join('; ');
}

/**
 * Validate an inbound server message. Never throws — transports call this on
 * every frame and a malformed frame must degrade to a logged warning rather
 * than tearing down the socket.
 */
export function parseServerMessage(raw: unknown): ParseResult<ServerMessage> {
  const res = ServerMessageSchema.safeParse(raw);
  return res.success
    ? { ok: true, value: res.data }
    : { ok: false, error: formatIssues(res.error) };
}

/** Validate an outbound command before it hits the wire. */
export function parseCommand(raw: unknown): ParseResult<Command> {
  const res = CommandSchema.safeParse(raw);
  return res.success
    ? { ok: true, value: res.data }
    : { ok: false, error: formatIssues(res.error) };
}

/** All layer keys, in the order the Layers tab should present them. */
export const LAYER_KEYS = LayerKeySchema.options;
